### DISCUSSION TITLE
Roadmap for Arrhenius.jl

### DISCUSSION THREAD

**Discussion by jiweiqi** [DENG-MIT/Arrhenius.jl#68] [Ideas]
![image](https://user-images.githubusercontent.com/8445647/118387233-4c7af280-b5eb-11eb-9412-06fdc4a81420.png)

**Comment by longemen3000** [comment]
on **super-critical, real gas states and condensed phases**, check out our work on https://github.com/ypaul21/OpenSAFT.jl. basically, we have a lot of real EOS based on Helmholtz energy for direct use. in particular, GERG2008 EoS would be ideal for obtaining natural gas properties (enthalpy, entropy, cp, cv). there are also all standard cubics (with the ability to couple them with appropriate ideal Helmholtz energies fitted from Cp) and, as the package name says, SAFT EoS.

**Reply by jiweiqi** [reply]
@TJP-Karpowski might be interested in the package.

**Reply by TJP-Karpowski** [reply]
Hi @longemen3000, 
indeed I am interested in your package. I have seen it before, but assumed it was abandoned since the main was not updated for months. I missed, that the dev branch is quite active :). I have two questions about your package: 
Is it already possible to calculated partial quantities? For combustion simulations I need not only cp of the mixture but also the vector of cp_i for each component. If this is not implemented so far is your package differentiable? Then cp_i = d cp/dX_i could be used.

**Reply by longemen3000** [reply]
the package is entirely differentiable (it has to be, some equations are too large to be manually differentiated and too sensitive to use finite differences) for example, Then cp_i = d cp/dX_i could be written as:
```
using OpenSAFT, ForwardDiff
#supposing p-T-n basis
cp(model,p,T,n) = OpenSAFT.isobaric_heat_capacity(model,p,T,n)
dcpdx(model,p,T,n) = ForwardDiff.gradient(z->cp(model,p,T,z),n)
```
this would perform a volume calculation to transform to v-t-n, calculate the phase and return the phase with least gibbs, and perform the gradient at that point.

example of working code:
```
julia> model = GERG2008(["methane","hydrogen","oxygen","water","carbon dioxide","carbon monoxide"])      
6-element GERG008 model, with compounds:
 n1 : methane
 n2 : hydrogen
 n3 : oxygen
 n4 : water
 n5 : carbon dioxide
 n6 : carbon monoxide

julia> dcpdx(model,101255,373,[0.1,0.1,0.1,0.1,0.1,0.5])
6-element Vector{Float64}:
 -1.2747935981224246e-6
  1.3005896691675966e-5
  3.090717979214906e-6
 -1.9878983410279718e-5
 -9.434088141602852e-6
  2.8982500958244796e-6

julia> @btime dcpdx(model,101255,373,[0.1,0.1,0.1,0.1,0.1,0.5])       
  504.400 μs (22 allocations: 8.41 
KiB)
6-element Vector{Float64}:
 -1.2747935981224246e-6
  1.3005896691675966e-5
  3.090717979214906e-6
 -1.9878983410279718e-5
 -9.434088141602852e-6
  2.8982500958244796e-6
```
considering that GERG2008 is the method used in [ISO 20765-2](https://www.iso.org/standard/59222.html). it could be useful in case of requiring standard-aproved calculations

**Reply by longemen3000** [reply]
the cubic volume solver was not AD friendly (we use polynomial solvers here). now it is, so with a faster cubic model:
```julia
julia> model = SRK(["methane","hydrogen","oxygen","water","carbon dioxide","carbon monoxide"])
SRK{BasicIdeal} with 6 components:
 "methane"
 "hydrogen"
 "oxygen"
 "water"
 "carbon dioxide"
 "carbon monoxide"
Contains parameters: a, b, acentricfactor, Tc, Pc, Mw

julia> @btime dcpdx(model,101255,373,[0.1,0.1,0.1,0.1,0.1,0.5])
  11.500 μs (24 allocations: 12.27 KiB)
6-element Vector{Float64}:
 -9.063472362581034e-6
  4.9935275781942505e-6
 -4.198242876572115e-6
 -2.358166800331494e-5
 -1.8727960071363622e-5
 -3.0679237614560442e-6
```

**Reply by TJP-Karpowski** [reply]
Wow, thanks for the detailed answer. This looks very good :). I will try it.

**Comment by jiweiqi** [comment]
**electrochemistry**

References for electrochemistry in Cantera
[lithium_ion_battery in Cantera](https://cantera.org/examples/jupyter/electrochemistry/lithium_ion_battery.ipynb.html)
